pyflwdir/streams.py

```diff
@@ -308,3 +308,65 @@ def stream_distance(
             d = gis_utils.distance(idx0, idx_ds, ncol, latlon, transform)
         dist[idx0] = dist[idx_ds] + d
     return dist
+
+
+@njit
+def smooth_rivlen(
+    idxs_ds,
+    seq,
+    rivlen,
+    min_rivlen,
+    smooth_cells=3,
+    mask=None,
+):
+    """Return smoothed river length, by taking the average of river length
+    `rivlen` per branch (or a minimum river cells of `smooth_cells`) with a river
+    length <= `min_rivlen`.
+
+    Parameters
+    ----------
+    rivlen : 1D array
+        River length values.
+    min_rivlen : float
+        Minimum river length.
+    smooth_cells : int
+        Minimum number of river cells to smooth, by default 3.
+    mask : 1D array of boolean
+        Mask of river cells to consider.
+
+    Returns
+    -------
+    1D array of float
+        River length values.
+    """
+    mv = -9999.0
+    nup = core.upstream_count(idxs_ds=idxs_ds, mask=mask, mv=core._mv)
+    rivlen_out = np.full(idxs_ds.size, mv, dtype=np.float32)
+    done = np.array([bool(0) for _ in range(idxs_ds.size)])  # all False
+    cnt = 0
+    for idx0 in seq[::-1]:  # up- to downstream
+        if done[idx0] or (mask is not None and ~mask[idx0]):
+            continue
+        idxs = [idx0]  # initiate with correct dtype
+        while True:
+            idx_ds = idxs_ds[idx0]
+            pit = idx_ds == idx0
+            if not pit:
+                if not done[idx_ds]:
+                    idxs.append(idx_ds)
+            if nup[idx_ds] > 1 and len(idxs) > smooth_cells or pit:
+                if pit:
+                    l = len(idxs)
+                    inds = np.array(idxs, dtype=idxs_ds.dtype)
+                else:
+                    l = len(idxs) - 1
+                    inds = np.array(idxs[0:-1], dtype=idxs_ds.dtype)
```

Comment:
Because of this line `idxs[0:-1]` the confluence cell (with more than one upstream neighbor) is never considered. Perhaps we should work with stream order instead of number of upstream neighbors to fix this.